mailserver/mailserver_test.go

```diff
@@ -45,83 +47,312 @@ type ServerTestParams struct {
+		},
+		{
+			config:        *s.config,
+			expectedError: nil,
+			limiterActive: true,
+			info:          "Initializing a mail server with a config with correct config and active limiter",
+		},
+		{
+			config: params.WhisperConfig{
+				DataDir:  "/tmp/",
+				Password: "pwd",
+			},
+			expectedError: nil,
+			limiterActive: false,
+			info:          "Initializing a mail server with a config with empty DataDir and inactive limiter",
+		},
+	}
+
+	for _, tc := range testCases {
+		s.T().Run(tc.info, func(*testing.T) {
+			s.server.limit = nil
+			err := s.server.Init(s.shh, &tc.config)
+			s.server.tick = nil
+			s.server.Close()
+			s.Equal(tc.expectedError, err)
+			s.Equal(tc.limiterActive, (s.server.limit != nil))
+		})
 	}
 }
 
-func TestDBKey(t *testing.T) {
+func (s *MailserverSuite) TestArchive() {
+	err := s.server.Init(s.shh, s.config)
+	s.server.tick = nil
+	s.NoError(err)
+	defer s.server.Close()
+
+	env, err := generateEnvelope(time.Now())
+	s.NoError(err)
+	rawEnvelope, err := rlp.EncodeToBytes(env)
+	s.NoError(err)
+
+	s.server.Archive(env)
+	key := NewDbKey(env.Expiry-env.TTL, env.Hash())
+	archivedEnvelope, err := s.server.db.Get(key.raw, nil)
+	s.NoError(err)
+
+	s.Equal(rawEnvelope, archivedEnvelope)
+}
+
+func (s *MailserverSuite) TestManageLimits() {
+	s.server.limit = newLimiter(time.Duration(5) * time.Millisecond)
+	s.server.managePeerLimits([]byte("peerID"))
+	s.Equal(1, len(s.server.limit.db))
+	firstSaved := s.server.limit.db["peerID"]
+
+	// second call when limit is not accomplished does not store a new limit
+	s.server.managePeerLimits([]byte("peerID"))
+	s.Equal(1, len(s.server.limit.db))
+	s.Equal(firstSaved, s.server.limit.db["peerID"])
+}
+
+func (s *MailserverSuite) TestDBKey() {
 	var h common.Hash
 	i := uint32(time.Now().Unix())
 	k := NewDbKey(i, h)
-	assert(len(k.raw) == common.HashLength+4, "wrong DB key length", t)
-	assert(byte(i%0x100) == k.raw[3], "raw representation should be big endian", t)
-	assert(byte(i/0x1000000) == k.raw[0], "big endian expected", t)
+	s.Equal(len(k.raw), common.HashLength+4, "wrong DB key length")
+	s.Equal(byte(i%0x100), k.raw[3], "raw representation should be big endian")
+	s.Equal(byte(i/0x1000000), k.raw[0], "big endian expected")
 }
 
-func TestMailServer(t *testing.T) {
+func (s *MailserverSuite) TestMailServer() {
 	var server WMailServer
 
-	setupServer(t, &server)
+	s.setupServer(&server)
 	defer server.Close()
 
-	env := generateEnvelope(t, time.Now())
+	env, err := generateEnvelope(time.Now())
+	s.NoError(err)
+
 	server.Archive(env)
-	deliverTest(t, &server, env)
+	testCases := []struct {
+		params      *ServerTestParams
+		emptyLow    bool
+		lowModifier int32
+		uppModifier int32
+		topic       byte
+		expect      bool
+		shouldFail  bool
+		info        string
+	}{
+		{
+			params:      s.defaultServerParams(env),
+			lowModifier: 0,
+			uppModifier: 0,
+			expect:      true,
```

Comment:
This function has a few test cases but only this one expects some results to be returned. Moreover, almost all of these test cases focus on request params. Can we maybe move them to a separate function and simplify its body? 

Overall, the body is overcomplicated and with so many variables in the test case, it's hard to figure out on which part we focus here. Clearer separation between unit and integration tests could be a key to refactor it. Integration tests can be even written in a separate file in a different package, for example `package mailserver_test`.

Reply:
@adambabik not sure if its what you meant by `can we maybe move them to a separate function` but I've simplified its body like:
```go
                  ... 
		{
			params: func() *ServerTestParams {
				params := s.defaultServerParams(env)
				params.low = 0
				params.upp = params.birth + 24

				return params
			}(),
			isOK: false,
			info: "Processing a request where difference between from and to is > 24 should fail",
		},
	}
	for _, tc := range testCases {
		s.T().Run(tc.info, func(*testing.T) {
			request := s.createRequest(tc.params)
			src := crypto.FromECDSAPub(&tc.params.key.PublicKey)
			ok, lower, upper, bloom := s.server.validateRequest(src, request)
			s.Equal(tc.isOK, ok)
			if ok {
				s.Equal(tc.params.low, lower)
				s.Equal(tc.params.upp, upper)

				expectedBloom := whisper.TopicToBloom(tc.params.topic)
				s.Equal(expectedBloom, bloom)
				s.Equal(tc.expect, s.messageExists(env, tc.params.low, tc.params.upp, bloom))

				src[0]++
				ok, _, _, _ = s.server.validateRequest(src, request)
				s.True(ok)
			}
		})
	}
```
which it's much clear